[move] Implement generic comparison method in move cmp::compare

igor-aptos opened the pull request:
## Description
Implement generic comparison operator for any move type. 
This enables writing generic methods that require comparison - for example binary_search on a vector.

Since move doesn't have signed integers, comparison return value is not intuitive (increased by 1 from usual -1, 0, 1), and so is wrapped into Ordering struct - so users don't need to think about it (with move 2, calling those will be pretty intuitive)

## Type of Change
- [x] New feature

## Which Components or Systems Does This Change Impact?
- [x] Move/Aptos Virtual Machine

## How Has This Been Tested?
Provided unit tests

## Key Areas to Review

## Checklist
- [ ] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I identified and added all stakeholders and component owners affected by this change as reviewers
- [ ] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

<!-- Thank you for your contribution! -->

runtian-zhou commented on aptos-move/framework/aptos-stdlib/sources/cmp.move:
```diff
@@ -0,0 +1,90 @@
+    /// 2   iff first value is larger than the second
+    native fun compare_impl<T>(first: &T, second: &T): u8;
+
+    struct Ordering has copy, drop {
```
Can we use enum for this now? cc @wrwg

igor-aptos replied:
would enum be better here? Would enums be more inefficient here? 

how would the native function return an enum that is defined in the module here?

runtian-zhou commented on aptos-move/framework/aptos-stdlib/sources/cmp.move:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    // #[test]
+    // fun test_compare_enums() {
```
Why is enum ignored?

igor-aptos replied:
enum's are not yet enabled in the framework.

msmouse commented on third_party/move/move-vm/types/src/values/values_impl.rs:
```diff
@@ -543,6 +544,55 @@ impl ValueImpl {
+            | (IndexedRef(_), _)
+            | (DelayedFieldID { .. }, _) => {
+                return Err(PartialVMError::new(StatusCode::INTERNAL_TYPE_ERROR)
+                    .with_message(format!("cannot compare values: {:?}, {:?}", self, other)))
```
Shall we use `Display` instead of `Debug` in the message (hopefully that means more stable error message)?

igor-aptos replied:
this is copied from equals. I am not sure what is more stable from MoveVM prespective, but seems like this should be the same as in equals. 

and changing equals would require feature flag? if we wanted to change eeverything to {} , we should do that separately, I think

msmouse replied:
okay.. putting Debug text to messages we dare not to change generally make me very sad but you don't make it worse I guess

igor-aptos replied:
@vgao1996 for thoughts

wrwg replied:
The error message of VMError is (more or less unfortunately, making it hard to show to users) not stored on chain, so I do not think it is very important to be stable.

msmouse commented on third_party/move/move-vm/types/src/values/values_impl.rs:
```diff
@@ -595,12 +645,65 @@ impl Container {
+            | (VecAddress(_), _) => {
+                return Err(
+                    PartialVMError::new(StatusCode::INTERNAL_TYPE_ERROR).with_message(format!(
+                        "cannot compare container values: {:?}, {:?}",
```
ditto

lightmark commented on aptos-move/aptos-gas-schedule/src/gas_schedule/move_stdlib.rs:
```diff
@@ -36,5 +39,8 @@ crate::gas_schedule::macros::define_gas_parameters!(
         [bcs_serialized_size_per_byte_serialized: InternalGasPerByte, { RELEASE_V1_18.. => "bcs.serialized_size.per_byte_serialized" }, 36],
         [bcs_serialized_size_failure: InternalGas, { RELEASE_V1_18.. => "bcs.serialized_size.failure" }, 3676],
+
+        [cmp_compare_base: InternalGas, { RELEASE_V1_21.. => "cmp.base" }, 367],
```
how do we get 367?

igor-aptos replied:
copied from equality costs

lightmark commented on aptos-move/framework/move-stdlib/sources/cmp.move:
```diff
@@ -0,0 +1,105 @@
+module std::cmp {
+    const EQUAL: u8 = 1;
+    const LESS_THAN: u8 = 0;
+    const GREATER_THAN: u8 = 2;
```
why not enum?

lightmark commented on aptos-move/framework/move-stdlib/sources/cmp.move:
```diff
@@ -0,0 +1,105 @@
+    /// 2   iff first value is larger than the second
+    native fun compare_impl<T>(first: &T, second: &T): u8;
+
+    struct Ordering has copy, drop {
+        value: u8,
+    }
```
less_than, equal, greater_than can be ordering and return value.

igor-aptos commented:
I cannot use move 2 features (i.e. enums), until framework is compiled with it. 

I can modify it in additional commit in the stack, and decide when landing.

igor-aptos commented:
updated to use enums

wrwg commented on aptos-move/framework/aptos-stdlib/doc/cmp.md:
That is surprising -- why do we have this .md in the aptos-stdlib? Leftover?

wrwg commented on aptos-move/framework/move-stdlib/sources/cmp.move:
```diff
@@ -0,0 +1,98 @@
+        self is Ordering::GreaterThan
+    }
+
+    public fun is__greater_or_equal(self: &Ordering): bool {
```
s/__/_

wrwg commented on aptos-move/framework/move-stdlib/sources/cmp.move:
```diff
@@ -0,0 +1,98 @@
+        GreaterThan,
+    }
+
+    public native fun compare<T>(first: &T, second: &T): Ordering;
```
Please add a `///` doc comment to this function. You should likely outline that you do lexicographical ordering, with one or two examples.

wrwg commented on aptos-move/framework/move-stdlib/sources/cmp.move:
```diff
@@ -0,0 +1,98 @@
+
+    #[test_only]
+    enum SomeEnum has drop {
+        V1 { field_1: u64 },
```
For good test coverage, a nested enum might be helpful, and have a field with a vector. Also strings and signers would be interesting.

wrwg commented on aptos-move/framework/move-stdlib/sources/cmp.move:
```diff
@@ -0,0 +1,98 @@
+    fun test_compare_structs() {
+        assert!(compare(&SomeStruct { field_1: 1, field_2: 2}, &SomeStruct { field_1: 1, field_2: 3}) is Ordering::LessThan, 0);
+        assert!(compare(&SomeStruct { field_1: 1, field_2: 2}, &SomeStruct { field_1: 1, field_2: 1}) is Ordering::GreaterThan, 1);
+        assert!(compare(&SomeStruct { field_1: 1, field_2: 2}, &SomeStruct { field_1: 1, field_2: 1}) is Ordering::GreaterThan, 2);
```
Test Equal case?

ziaptos replied:
Ditto for the tests below.

wrwg commented on aptos-move/framework/move-stdlib/src/natives/cmp.rs:
```diff
@@ -0,0 +1,79 @@
+// from that (b) assuming that `vector<u8>` embedded in a string
+// is already valid utf8, we can use `str::from_utf8_unchecked` to
+// create a `&str` view on the bytes without a copy. Once we have this
+// view, we can call ut8 functions like length, substring, etc.
```
Is that a left over C&P from the string module? Sounds familiar...

wrwg commented on aptos-move/framework/move-stdlib/sources/cmp.move:
```diff
@@ -0,0 +1,98 @@
+module std::cmp {
+    const EQUAL: u8 = 1;
```
Why those constants and enums?

ziaptos replied:
Curious as well

wrwg commented on aptos-move/framework/move-stdlib/src/natives/cmp.rs:
```diff
@@ -0,0 +1,79 @@
+const ORDERING_GREATER_THAN_VARIANT: u16 = 2;
+
+/***************************************************************************************************
+ * native fun internal_check_utf8
```
Forgot to update

wrwg commented on aptos-move/framework/move-stdlib/src/natives/cmp.rs:
```diff
@@ -0,0 +1,79 @@
+/***************************************************************************************************
+ * native fun internal_check_utf8
+ *
+ *   gas cost: base_cost + unit_cost * length_in_bytes
```
Probably then this one too.

wrwg commented on third_party/move/move-vm/types/src/values/values_impl.rs:
```diff
@@ -543,6 +544,55 @@ impl ValueImpl {
+            // semantics is that identifiers might not be deterministic, and
+            // therefore equality can have different outcomes on different nodes
+            // of the network. Note that the error returned here is not an
+            // invariant violation but a runtime error.
```
Does equality already do the same? 

I wonder whether this should use the same error code as equality, which definitely is not VM_EXTENSION__ERROR.

ziaptos replied:
Comment copy-pasted from fn equals. Let's at least change Disallow equality to Disallow comparison. I'd also remove rest of the comment and just write (see ValueImpl::fn equals).

@wrwg The VM_EXTENSION_ERROR makes sense, it's an error of the aggregators extension.

wrwg commented on third_party/move/move-vm/types/src/values/values_impl.rs:
```diff
@@ -543,6 +544,55 @@ impl ValueImpl {
+            | (ContainerRef(_), _)
+            | (IndexedRef(_), _)
+            | (DelayedFieldID { .. }, _) => {
+                return Err(PartialVMError::new(StatusCode::INTERNAL_TYPE_ERROR)
```
The error message should not be 'cannot compare values' but more like 'inconsistent types for arguments of comparison" or similar. Should never happen with paranoid check, but nevertheless.

ziaptos commented on aptos-move/framework/move-stdlib/sources/cmp.move:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    #[test]
+    fun test_compare_vectors() {
```
`GreaterThan` and `Equal` missing

ziaptos commented on aptos-move/framework/move-stdlib/sources/cmp.move:
```diff
@@ -0,0 +1,98 @@
+
+    enum Ordering has copy, drop {
+        /// First value is less than the second value.
+        LessThan,
```
I'd strongly prefer to follow rust here, and have these be `Less`, `Greater` and `Equal`

igor-aptos commented on aptos-move/framework/aptos-stdlib/doc/cmp.md:
I've modeled by rust, just a longer version:

`is_lt` vs `is_le`

https://doc.rust-lang.org/std/cmp/enum.Ordering.html

@ziaptos given that, what do you prefer? Should I change to shorthands (i.e. same as rust), or to one of the longer variants?

ziaptos replied:
I'd prefer rust short ones.

igor-aptos commented:
addressed comments and added tests

showed also that doing bcs::to_bytes and comparing would produce different and not natural ordering (due to little-endian serialization), showcasing the need for cmp::compare for primitive types as well, not just for performance.

igor-aptos commented:
moved move files with enums into separate PR (#15245), so this can land without waiting on move2

runtian-zhou reviewed (changes requested):
I would aim for 100% code coverage given the number of matches in the `value_impl.rs`. Should we land the move file first as a private function that is not called elsewhere just so we can add the tests first?

runtian-zhou commented on third_party/move/move-vm/types/src/values/values_impl.rs:
```diff
@@ -704,12 +812,117 @@ impl IndexedRef {
+    fn compare(&self, other: &Self) -> PartialVMResult<Ordering> {
+        use Container::*;
+
+        let res = match (
```
Was wondering if the logic would be easier if you do a read_ref on both ends and compare?

igor-aptos replied:
not sure, why equals doesn't do that? maybe perf?

runtian-zhou commented on third_party/move/move-vm/types/src/values/values_impl.rs:
```diff
@@ -536,8 +537,62 @@ impl ValueImpl {
+            (Bool(l), Bool(r)) => l.cmp(r),
+            (Address(l), Address(r)) => l.cmp(r),
+
+            (Container(l), Container(r)) => l.compare(r)?,
```
Can a `Container` be compared with `IndexedRef` etc?

igor-aptos replied:
no, as equals doesn't have those either.

igor-aptos commented:
@runtian-zhou do you mean 100% code coverage with move unit tests, or something else?

For the complexity of code - code (i.e. all the matches) have been copied 1-1 from equals, so they should cover all the cases exactly.

signature of this move function uses enums, so I cannot land the move file with private function, until move 2 is enabled for the framework. So I've separated them out - but will confirm all tests in #15245 pass, before I land this one.

ziaptos reviewed (approved):
I've reviewed this before, but didn't stamp
